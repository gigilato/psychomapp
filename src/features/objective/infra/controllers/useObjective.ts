import { useQuery } from '@tanstack/react-query'

import { getObjective } from '../fetchers'

export const getQueryKey = (id: string) => ['objectives', id]
export const useObjective = (id?: string) => {
  const query = useQuery({
    queryKey: getQueryKey(id!),
    queryFn: async () => {
      return getObjective(id!)
    },
    enabled: !!id,
  })
  return query
}
