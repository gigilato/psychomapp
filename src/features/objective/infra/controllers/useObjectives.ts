import { useQuery } from '@tanstack/react-query'

import { getObjectives } from '../fetchers'

export const queryKey = ['objectives']
export const useObjectives = () => {
  const query = useQuery({
    queryKey,
    queryFn: async () => {
      return getObjectives()
    },
  })
  return query
}
