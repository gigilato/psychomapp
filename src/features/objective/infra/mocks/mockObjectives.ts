import { Objective } from '$types/database'

export const mockObjectiveList: Objective[] = [
  {
    id: '1',
    name: 'Obj1',
    slug: 'Obj1',
    color: '#1fff00',
    createdAt: '',
    updatedAt: '',
    practitionnerId: '',
  },
  {
    id: '2',
    name: 'Obj2',
    slug: 'Obj2',
    color: '#ff3f00',
    createdAt: '',
    updatedAt: '',
    practitionnerId: '',
  },
  {
    id: '3',
    name: 'Obj3',
    slug: 'Obj3',
    color: '#ffff00',
    createdAt: '',
    updatedAt: '',
    practitionnerId: '',
  },
  {
    id: '4',
    name: 'Obj4',
    slug: 'Obj4',
    color: '#100f50',
    createdAt: '',
    updatedAt: '',
    practitionnerId: '',
  },
]
