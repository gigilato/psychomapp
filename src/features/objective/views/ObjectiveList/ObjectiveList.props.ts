import { ReactElement } from 'react'

import { IBoxProps } from '$atoms'
import { Objective } from '$types/database'

export interface IObjectiveListProps extends IBoxProps {
  onPressObjective?: (objective: Objective) => void
  renderError?: () => ReactElement
  renderEmpty?: () => ReactElement
}

export interface IObjectiveListContainerProps extends IObjectiveListProps {
  data?: Objective[]
  isLoading?: boolean
  error?: Error | null
}
