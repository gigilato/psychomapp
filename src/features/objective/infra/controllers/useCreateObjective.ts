import { useMutation, useQueryClient } from '@tanstack/react-query'

import { i18n } from '$infra/i18n'
import { showToast } from '$infra/toast'

import { queryKey } from './useObjectives'
import { CreateObjectiveVariables, createObjective, getObjective } from '../fetchers'

export const useCreateObjective = () => {
  const queryClient = useQueryClient()

  const mutation = useMutation({
    mutationFn: async (variables: CreateObjectiveVariables) => {
      const { result, data } = await createObjective(variables)
      if (result.status !== 201) throw new Error('CreateObjectiveError')
      const objective = await getObjective(data.id)
      return objective
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey })
    },
    onError: () => {
      showToast(i18n.t('errors.default'))
    },
  })
  return mutation
}
