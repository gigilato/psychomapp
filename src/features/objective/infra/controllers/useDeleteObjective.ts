import { useMutation, useQueryClient } from '@tanstack/react-query'

import { i18n } from '$infra/i18n'
import { showToast } from '$infra/toast'
import { Objective } from '$types/database'

import { queryKey } from './useObjectives'
import { DeleteObjectiveVariables, deleteObjective } from '../fetchers'

export const useDeleteObjective = () => {
  const queryClient = useQueryClient()

  const mutation = useMutation({
    mutationFn: async (variables: DeleteObjectiveVariables) => {
      const result = await deleteObjective(variables.id)
      if (result.status !== 204) throw new Error('DeleteObjectiveError')
      return variables
    },
    onMutate: async (variables) => {
      await queryClient.cancelQueries({ queryKey })
      const previous = queryClient.getQueryData<Objective[]>(queryKey)
      queryClient.setQueryData<Objective[]>(
        queryKey,
        (current) => current?.filter((item) => item.id !== variables.id)
      )
      return { previous }
    },
    onError: (_, variables, context) => {
      if (context?.previous) queryClient.setQueryData<Objective[]>(queryKey, context.previous)
      showToast(i18n.t('errors.default'))
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey })
    },
  })
  return mutation
}
