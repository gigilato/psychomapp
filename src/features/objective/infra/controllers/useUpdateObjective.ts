import { useMutation, useQueryClient } from '@tanstack/react-query'

import { i18n } from '$infra/i18n'
import { showToast } from '$infra/toast'
import { Objective } from '$types/database'

import { queryKey } from './useObjectives'
import { UpdateObjectiveVariables, updateObjective } from '../fetchers'

export const useUpdateObjective = () => {
  const queryClient = useQueryClient()

  const mutation = useMutation({
    mutationFn: async (variables: UpdateObjectiveVariables) => {
      const { result, data } = await updateObjective(variables)
      if (result.status !== 204) throw new Error('UpdateObjectiveError')
      return data
    },
    onMutate: async ({ current, update }) => {
      await queryClient.cancelQueries({ queryKey })
      const previous = queryClient.getQueryData<Objective[]>(queryKey)
      if (!previous) return
      const updated = previous?.map((item) => {
        if (item.id === current.id) {
          return { ...item, ...update }
        }
        return item
      })
      queryClient.setQueryData<Objective[]>(queryKey, updated)
      return { previous }
    },
    onError: (_, variables, context) => {
      if (context?.previous) queryClient.setQueryData<Objective[]>(queryKey, context.previous)
      showToast(i18n.t('errors.default'))
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey })
    },
  })
  return mutation
}
