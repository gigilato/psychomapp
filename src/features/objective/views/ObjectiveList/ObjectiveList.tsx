import { Skeleton } from 'moti/skeleton'

import { Box, HStack, Text } from '$atoms'
import { useObjectives } from '$features/objective/infra/controllers/useObjectives'
import { i18n } from '$infra/i18n'
import { Tag, TagContainer } from '$molecules'
import { ErrorState } from '$organisms'

import { IObjectiveListProps, IObjectiveListContainerProps } from './ObjectiveList.props'

export const ObjectiveListContainer = ({
  className,
  onPressObjective,
  data,
  isLoading: _isLoading,
  error,
  renderEmpty = () => <Text variant="smallBody">{i18n.t('objectives.empty')}</Text>,
  renderError = () => <Text variant="smallBody">{i18n.t('errors.default')}</Text>,
  ...props
}: IObjectiveListContainerProps) => {
  const isLoading = !data && _isLoading

  if (isLoading)
    return (
      <Skeleton.Group show={isLoading}>
        <HStack className={className} {...props}>
          <Skeleton colorMode="light">
            <Box className="h-s w-[30]" />
          </Skeleton>
          <Box className="w-xxs" />
          <Skeleton colorMode="light">
            <Box className="h-s w-[50]" />
          </Skeleton>
          <Box className="w-xxs" />
          <Skeleton colorMode="light">
            <Box className="h-s w-[40]" />
          </Skeleton>
        </HStack>
      </Skeleton.Group>
    )
  if (!data && error)
    return (
      <Box className={className} {...props}>
        {renderError()}
      </Box>
    )
  if (data?.length === 0)
    return (
      <Box className={className} {...props}>
        {renderEmpty()}
      </Box>
    )

  return (
    <Box className={className} {...props}>
      <TagContainer gap={8}>
        {data?.map((objective) => (
          <Tag
            key={objective.id}
            tag={objective.name}
            backgroundColor={objective.color}
            onPress={onPressObjective ? () => onPressObjective(objective) : undefined}
          />
        ))}
      </TagContainer>
    </Box>
  )
}
export const ObjectiveList = (props: IObjectiveListProps) => {
  const { data, isLoading, error } = useObjectives()
  return (
    <ObjectiveListContainer
      data={data}
      isLoading={isLoading}
      error={error}
      renderError={() => <ErrorState width={100} />}
      renderEmpty={() => <ErrorState type="empty" width={100} />}
      {...props}
    />
  )
}
