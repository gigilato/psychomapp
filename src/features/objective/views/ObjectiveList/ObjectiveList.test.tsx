import { render } from '@testing-library/react-native'

import { useObjectives } from '$features/objective/infra/controllers/useObjectives'
import { mockObjectiveList } from '$features/objective/infra/mocks/mockObjectives'
import { QueryWrapper } from '$jest/jest.wrappers'

import { ObjectiveList } from './ObjectiveList'

jest.mock('$features/objective/infra/controllers/useObjectives', () => ({
  useObjectives: jest.fn(),
}))

describe('ObjectiveList', () => {
  it('should render correctly', async () => {
    ;(useObjectives as jest.Mock).mockImplementation(() => ({
      data: mockObjectiveList,
      isLoading: false,
      error: false,
    }))

    const Component = render(<ObjectiveList />, { wrapper: QueryWrapper })

    expect(Component).toMatchSnapshot()
  })

  it('should render loading', async () => {
    ;(useObjectives as jest.Mock).mockImplementation(() => ({
      data: undefined,
      isLoading: true,
      error: false,
    }))

    const Component = render(<ObjectiveList />, { wrapper: QueryWrapper })

    expect(Component).toMatchSnapshot()
  })

  it('should render error', async () => {
    ;(useObjectives as jest.Mock).mockImplementation(() => ({
      data: undefined,
      isLoading: false,
      error: true,
    }))

    const Component = render(<ObjectiveList />, { wrapper: QueryWrapper })

    expect(Component).toMatchSnapshot()
  })
})
