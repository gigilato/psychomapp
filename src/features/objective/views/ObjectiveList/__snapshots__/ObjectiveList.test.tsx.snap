// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectiveList should render correctly 1`] = `
<View
  style={
    [
      [
        undefined,
        undefined,
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "flexWrap": "wrap",
        },
        [
          undefined,
          [
            {
              "marginLeft": -8,
              "marginTop": -8,
            },
            undefined,
          ],
        ],
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderBottomLeftRadius": 9999,
            "borderBottomRightRadius": 9999,
            "borderTopLeftRadius": 9999,
            "borderTopRightRadius": 9999,
          },
          {
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          },
          {
            "flexDirection": "row",
          },
          {
            "alignSelf": "flex-start",
          },
          [
            {
              "opacity": undefined,
            },
            [
              {
                "backgroundColor": "#1fff00",
              },
              {
                "marginLeft": 8,
                "marginTop": 8,
              },
              undefined,
            ],
          ],
        ]
      }
      testID="tag/Obj1"
    >
      <Text
        style={
          [
            {
              "fontFamily": "SpaceGrotesk-light",
            },
            {
              "fontSize": 12,
            },
            {
              "lineHeight": 16,
            },
            {
              "color": "#FFFCF9",
            },
          ]
        }
      >
        Obj1
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderBottomLeftRadius": 9999,
            "borderBottomRightRadius": 9999,
            "borderTopLeftRadius": 9999,
            "borderTopRightRadius": 9999,
          },
          {
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          },
          {
            "flexDirection": "row",
          },
          {
            "alignSelf": "flex-start",
          },
          [
            {
              "opacity": undefined,
            },
            [
              {
                "backgroundColor": "#ff3f00",
              },
              {
                "marginLeft": 8,
                "marginTop": 8,
              },
              undefined,
            ],
          ],
        ]
      }
      testID="tag/Obj2"
    >
      <Text
        style={
          [
            {
              "fontFamily": "SpaceGrotesk-light",
            },
            {
              "fontSize": 12,
            },
            {
              "lineHeight": 16,
            },
            {
              "color": "#FFFCF9",
            },
          ]
        }
      >
        Obj2
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderBottomLeftRadius": 9999,
            "borderBottomRightRadius": 9999,
            "borderTopLeftRadius": 9999,
            "borderTopRightRadius": 9999,
          },
          {
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          },
          {
            "flexDirection": "row",
          },
          {
            "alignSelf": "flex-start",
          },
          [
            {
              "opacity": undefined,
            },
            [
              {
                "backgroundColor": "#ffff00",
              },
              {
                "marginLeft": 8,
                "marginTop": 8,
              },
              undefined,
            ],
          ],
        ]
      }
      testID="tag/Obj3"
    >
      <Text
        style={
          [
            {
              "fontFamily": "SpaceGrotesk-light",
            },
            {
              "fontSize": 12,
            },
            {
              "lineHeight": 16,
            },
            {
              "color": "#FFFCF9",
            },
          ]
        }
      >
        Obj3
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderBottomLeftRadius": 9999,
            "borderBottomRightRadius": 9999,
            "borderTopLeftRadius": 9999,
            "borderTopRightRadius": 9999,
          },
          {
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
          },
          {
            "flexDirection": "row",
          },
          {
            "alignSelf": "flex-start",
          },
          [
            {
              "opacity": undefined,
            },
            [
              {
                "backgroundColor": "#100f50",
              },
              {
                "marginLeft": 8,
                "marginTop": 8,
              },
              undefined,
            ],
          ],
        ]
      }
      testID="tag/Obj4"
    >
      <Text
        style={
          [
            {
              "fontFamily": "SpaceGrotesk-light",
            },
            {
              "fontSize": 12,
            },
            {
              "lineHeight": 16,
            },
            {
              "color": "#FFFCF9",
            },
          ]
        }
      >
        Obj4
      </Text>
    </View>
  </View>
</View>
`;

exports[`ObjectiveList should render error 1`] = `
<View
  style={
    [
      [
        undefined,
        undefined,
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
        },
        {
          "alignSelf": "center",
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  >
    <ViewManagerAdapter_ExpoImage
      contentFit="contain"
      contentPosition={
        {
          "left": "50%",
          "top": "50%",
        }
      }
      height={100}
      onError={[Function]}
      onLoad={[Function]}
      onLoadStart={[Function]}
      onProgress={[Function]}
      placeholder={[]}
      source={[]}
      style={
        {
          "height": 100,
          "width": 100,
        }
      }
      transition={null}
      width={100}
    />
    <Text
      style={
        [
          {
            "fontFamily": "SpaceGrotesk-Regular",
          },
          {
            "fontSize": 16,
          },
          {
            "lineHeight": 18,
          },
          {
            "marginTop": 24,
          },
        ]
      }
    >
      Une erreur s'est produite
    </Text>
  </View>
</View>
`;

exports[`ObjectiveList should render loading 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      [
        undefined,
        undefined,
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "height": 12,
        },
        {
          "width": 30,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "width": 4,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "height": 12,
        },
        {
          "width": 50,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "width": 4,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "height": 12,
        },
        {
          "width": 40,
        },
        [
          undefined,
          undefined,
        ],
      ]
    }
  />
</View>
`;
